presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,308 @@
+    {
+        SourceDescriptor sourceDescriptor = SourceDescriptor.of(queryId, source);
+        DynamicFilterSummaryWithSenders dynamicFilterSummaryWithSenders = dynamicFilterSummaries.computeIfAbsent(sourceDescriptor,
+                ignored -> new DynamicFilterSummaryWithSenders());
+        DynamicFilterSummary mergedSummary = dynamicFilterSummaryWithSenders.addSummary(DynamicFilterSummaryWithSenders.StageTaskKey.of(stageId, taskId), driverId, dynamicFilterSummary, expectedSummariesCount);
+
+        if (mergedSummary != null) {
+            dynamicFilterSummaryWithSenders.getSummaryIfReady()
+                    .ifPresent(summary -> {
+                        SettableFuture<DynamicFilterSummary> future = futures.computeIfAbsent(sourceDescriptor, ignored -> SettableFuture.create());
+                        future.set(summary);
+                    });
+        }
+    }
+
+    /*
+     * registerTasks is required to know whether we have received dynamic filter summaries from all the tasks of query, after which the summary is ready to use
+     * Build side may start reporting summaries through storeOrMergeSummary before registerTasks runs
+     */
+    public void registerTasks(String source, Set<TaskId> taskIds)
+    {
+        if (taskIds.isEmpty()) {
+            return;
+        }
+
+        String queryId = taskIds.iterator().next().getQueryId().getId();
+        log.info("registerTasks source " + source + ", taskIds " + taskIds.stream().map(TaskId::toString).collect(Collectors.joining(",")));
+        checkArgument(taskIds.stream().allMatch(taskId -> taskId.getQueryId().getId().equals(queryId)), "All tasks have to belong to the same query");
+
+        DynamicFilterSummaryWithSenders dynamicFilterSummaryWithSenders = dynamicFilterSummaries.computeIfAbsent(SourceDescriptor.of(queryId, source),
+                ignored -> new DynamicFilterSummaryWithSenders());
+        dynamicFilterSummaryWithSenders.registerTasks(taskIds.stream()
+                .map(DynamicFilterSummaryWithSenders.StageTaskKey::of)
+                .collect(toImmutableSet()));
+    }
+
+    public ListenableFuture<DynamicFilterSummary> getSummary(String queryId, String source)
+    {
+        SourceDescriptor sourceDescriptor = SourceDescriptor.of(queryId, source);
+        SettableFuture<DynamicFilterSummary> future = futures.computeIfAbsent(sourceDescriptor, ignored -> SettableFuture.create());
+        DynamicFilterSummaryWithSenders dynamicFilterSummaryWithSenders = dynamicFilterSummaries.get(sourceDescriptor);
+        if (dynamicFilterSummaryWithSenders != null) {
+            dynamicFilterSummaryWithSenders.getSummaryIfReady()
+                    .ifPresent(future::set);
+        }
+        return future;
+    }
+
+    public void removeQuery(String queryId)
+    {
+        dynamicFilterSummaries.keySet().removeIf(sourceDescriptor -> sourceDescriptor.getQueryId().equals(queryId));
```

Comment:
`removeQuery` is called after completion of query so ideally we shouldn't see store summary calls coming after query completion. In case this does happen, query execution itself is unaffected but we would leave behind an extra entry in the map which wouldn't be removed.
The way to prevent this would be to ignore store summary calls from completed queries.
How about we maintain a set of recently completed query ids (last 10 minutes) and use that to ignore any late store summary calls ?

Reply:
> How about we maintain a set of recently completed query ids (last 10 minutes) and use that to ignore any late store summary calls

Sounds good to me.
@sopel39 WDYT?

Reply:
I think `DynamicFilterService` should pull dynamic filters itself from query info periodically. Then `storeSummary` won't be needed anymore and this class would become simpler as `HttpRemoteTask.java` won't push information into it.

Reply:
If we really need some notifications when dynamic filtering might be ready, then we should add:
`io.prestosql.execution.SqlQueryExecution#addQueryInfoListener`